### PR TITLE
Topic edit modal

### PR DESCRIPTION
**Pull request by dootMaster**
This draft PR is ready for review. What should a unit test for this feature entail?

### CODE REVIEW

**Comment by dootMaster** [comment]
https://user-images.githubusercontent.com/69024992/192124904-008e897c-b069-4129-8cda-36508ef26a49.mov

**Comment by dootMaster** [comment]
@chrisbobbe 👋

**Comment by dootMaster** [comment]
I think we also want the Edit Topic button to only appear for admins for now. I can borrow this:
```js
  if (roleIsAtLeast(ownUserRole, Role.Admin)) {
    buttons.push(deleteTopic);
  }
  ```

File: src/topics/TopicEditModal.js [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+      transparent
+      visible={topicModalState.visible}
+      animationType="slide"
+      onRequestClose={() => {
+        closeEditTopicModal();
+      }}
+    >
+      <View style={styles.wrapper}>
+        <View style={styles.modalView}>
+          <Text style={styles.titleText}>Edit topic</Text>
```
Comment (chrisbobbe) [inline src/topics/TopicEditModal.js:134]: Please use `ZulipTextIntl` instead of RN's plain `Text` component.

File: src/topics/TopicEditModal.js [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+            <Pressable
+              style={{
+                opacity: !topicState ? 0.25 : 1,
+                backgroundColor: BRAND_COLOR,
+                ...styles.button,
+              }}
+              onPress={handleSubmit}
+              disabled={!topicState}
+            >
+              <Text style={{ ...styles.text, ...styles.submitButtonText }}>Submit</Text>
```
Comment (chrisbobbe) [inline src/topics/TopicEditModal.js:165]: Please add "Submit" to static/translations/messages_en.json.

**Comment by chrisbobbe** [comment]
> I think we also want the Edit Topic button to only appear for admins for now. I can borrow this:
> 
> ```js
>   if (roleIsAtLeast(ownUserRole, Role.Admin)) {
>     buttons.push(deleteTopic);
>   }
> ```

That sounds right; please do! 🙂 Since it won't be obvious to everyone reading that code, please include a code comment saying why we're only showing it to admins-and-above, linking to discussion if any.

File: src/boot/TopicModalProvider.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    topic: '',
+    fetchArgs: {
+      auth: null,
+      messageId: null,
+      zulipFeatureLevel: null,
+    },
+  });
+
+  const startEditTopic = useCallback(
+    async (streamId, topic, streamsById, _) => {
```
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply src/boot/TopicModalProvider.js:44]: Also, let's have `startEditTopic` bail out (`return`) early if there's already an edit-topic interaction in progress, so we don't have two interactions stepping on each other and corrupting each other's state.

File: src/boot/TopicModalProvider.js [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+      closeEditTopicModal,
+    }),
+    [startEditTopic, closeEditTopicModal],
+  );
+
+  return (
+    <TopicModal.Provider value={topicModalHandler}>
+      {topicModalState.visible && (
+        <TopicEditModal topicModalState={topicModalState} topicModalHandler={topicModalHandler} />
+      )}
```
Comment (chrisbobbe) [inline src/boot/TopicModalProvider.js:88]: Hmm, is there a reason you'd like to unmount `TopicEditModal` when the modal isn't visible? As I understand it, RN's `Modal` component asks for a [`visible` prop](https://reactnative.dev/docs/modal#visible) (which we're ready already passing `topicModalState.visible` for) to specify that the modal isn't visible, and it's probably best to stick to that documented interface unless we have a reason not to. (I don't know if we risk things like breaking the exit animation with the current code.)

File: src/topics/TopicEditModal.js [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+type TopicEditModalArgs = {
+  topicModalState: {
+    visible: boolean,
+    topic: string,
+    fetchArgs: {
+      auth: Auth,
+      messageId: number,
+      zulipFeatureLevel: number,
+    },
```
Comment (chrisbobbe) [inline src/topics/TopicEditModal.js:17]: I think we don't need to set off this `fetchArgs` property. Is "fetch" a reference to the Fetch API, which we invoke deep down in the `apiFetch` function in src/api/apiFetch.js? If so, there are other things that make their way into that `fetch` call but don't appear here, like this endpoint's `send_notification_to_old_thread` param.

I see we have an existing name `fetchArgs` in the codebase, but I think that's meant to refer to the act of fetching data (messages) from the server. The main thing being done in the edit-topic UI is giving a command _to_ the server, to update its data.

**Review by chrisbobbe — COMMENTED** [review]
Thanks, @dootMaster! See comments below.

For the branch structure, I think it makes sense to have one NFC commit that converts `SearchMessagesCard` to a function component, followed by one non-NFC commit whose "single idea" is something like "add a new edit-topic UI" and contains all the work that accomplishes that. The added code is pretty coherent and is all meant to work together, so I think it's easier to see how it all works together when it's added in the same commit. Also please remember that all commits should pass all tests, and feel free to ask in `#git help` if you have questions on [fixing](https://zulip.readthedocs.io/en/latest/git/fixing-commits.html)/[tidying](https://zulip.readthedocs.io/en/latest/git/using.html#examine-and-tidy-your-commit-history) commits. 🙂

**Comment by dootMaster** [comment]
@chrisbobbe Thanks for reviewing! How is this achieved:

> Also, let's have startEditTopic bail out (return) early if there's already an edit-topic interaction in progress, so we don't have two interactions stepping on each other and corrupting each other's state.

File: src/boot/TopicModalProvider.js (hunk 4 above)
Comment (dootMaster) [reply src/boot/TopicModalProvider.js:88]: I was running into a problem where the initial topic string either wasn't displaying or wasn't resetting to the initial prop value after close. Thanks for poking at this! I've figured out that `useEffect` is my friend here.

File: src/boot/TopicModalProvider.js [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+  const startEditTopic = useCallback(
+    async (streamId, topic, streamsById, _) => {
+      const messageId = await fetchSomeMessageIdForConversation(
+        auth,
+        streamId,
+        topic,
+        streamsById,
+        zulipFeatureLevel,
+      );
```
[the remark this answers is not in the document]
Comment (dootMaster) [reply src/boot/TopicModalProvider.js:51]: Makes sense! This one in particular has me facepalming. 🤦

File: src/topics/TopicEditModal.js [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+  },
+  topicModalHandler: {
+    startEditTopic: (
+      streamId: number,
+      topic: string,
+      streamsById: Map<number, Stream>,
+      _: GetText,
+    ) => Promise<void>,
+    closeEditTopicModal: () => void,
+  },
```
[the remark this answers is not in the document]
Comment (dootMaster) [reply src/topics/TopicEditModal.js:27]: I've gotten rid of the handler container altogether, and opted to just pass the callbacks. My instinct to place them in a container comes from something else I worked on where I needed to add more to the handler and I regretted not using a container, but here it's likely that we won't be calling `startEditTopic` from the modal and it won't be seeing any functional changes. 😄

**Comment by chrisbobbe** [comment]
(For the record, the next revision of this PR went to #5510.)

**Pull request closed without merging**